### PR TITLE
Fix hash join with outer joins and zero/NULL values

### LINKED ISSUE #1214: Semi-Hash-Join mistreats NULL values
```
(debug)> print t
=== Columns
|       a||        MVCC        |
|     int||_BEGIN|_END  |_TID  |
=== Chunk 0 ===
|       0||     0|      |      |
|    NULL||     0|      |      |
|     123||     0|      |      |
|    1234||     0|      |      |
|   12345||     0|      |      |
(debug)> SELECT * FROM t t1 LEFT JOIN t t2 ON t1.a = t2.a
=== Columns
|       a|       a|
|     int|     int|
=== Chunk 0 ===
|       0|       0|
|    NULL|       0|   <<< What is this guy doing here?
|     123|     123|
|    1234|    1234|
|   12345|   12345|
===
5 rows total
```

My guess is that is has something to do with how values are materialized:

https://github.com/hyrise/hyrise/blob/6c0e75383cfe53cd25bc096eafe641ed87a095c4/src/lib/operators/join_hash/join_hash_steps.hpp#L121_L132

If a value is null but should be kept, we simply write the default value of the respective type and forget that the entry was null.

That should be fixed and we should add default values to most test files. We did not catch this because we always tested the joins with values like "123", but never "0".

### PR DESCRIPTION
**Pull request by Bouncner**
This PR fixes #1214.

Some of the changes:
* Making NULL consideration a template parameter (and that's something I'd like to get feedback on)
* Add a test that performs an outer join on a column with zero and NULL values. This test would fail in the current master.
* We now longer store the hash values. I have some veeery rough number of string joins and it makes a difference here, but it's not enough to keep the hashed value around for now.
* Cleaning up some tests (we now test join operator and the hash step functions separately)

Apart from that I had an issue with Clang and "invalid explicitly-specified argument" (296@join_hash.cpp). If anybody has an idea to solve that, please tell me.

Last but not least: performance does improve (should be mainly the removed hash values). I'll run TPC-H over night to measure the impact.

### CODE REVIEW

File: src/lib/operators/join_hash.cpp [hunk 1]
```diff
@@ -281,16 +263,32 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
-      // Materialize right table. 'keep_nulls' makes sure that the relation on
-      // the right materializes NULL values when executing an OUTER join.
-      materialized_right = materialize_input<RightType, HashedType>(right_in_table, _column_ids.second,
-                                                                    histograms_right, _radix_bits, keep_nulls);
+      // Materialize right table. The third template parameter signals if the relation on the right (probe
+      // relation) materializes NULL values when executing OUTER joins (default is to discard NULL values).
+      /*
+      TODO (anyone):
+        Clang 6.0.1 does not allow "materialize_input<RightType, HashedType, static_cast<bool>(keep_nulls)>" (error:
+        invalid explicitly-specified argument), which works on GCC (see https://stackoverflow.com/q/33872039/1147726).
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: Wait, what? Why does gcc accept that? I thought template arguments would have to be constexpr. `keep_nulls` obviously is not. Generally, it would be news to me if that would work: https://godbolt.org/z/xTZ6jc

The difference to the SO post is that they are using a constexpr expression as the template parameter.

If you want to, try and isolate a piece of code with a runtime value that gcc accepts but clang doesn't.

File: src/lib/operators/join_hash.cpp [hunk 2]
```diff
@@ -281,16 +263,32 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
+      */
+      if (keep_nulls) {
+        materialized_right = materialize_input<RightType, HashedType, true>(right_in_table, _column_ids.second,
+                                                                            histograms_right, _radix_bits);
+      } else {
+        materialized_right =
+            materialize_input<RightType, HashedType>(right_in_table, _column_ids.second, histograms_right, _radix_bits);
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: Not a fan of binary default template arguments. This should be `<..., false>`.

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 3]
```diff
@@ -70,15 +68,35 @@ template <typename T>
...
+
+  size_t offset = 0;
+  for (ChunkID chunk_id{0}; chunk_id < chunk_count; ++chunk_id) {
+    chunk_offsets[chunk_id] = offset;
+    offset += table->get_chunk(chunk_id)->size();
+  }
+  return chunk_offsets;
+}
+
+template <typename T, typename HashedType, bool consider_null_values = false>
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: Did you check if this has any performance implications? Remember that using this as a template parameter as opposed to a regular parameter doubles the compile time.

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 4]
```diff
@@ -70,15 +68,35 @@ template <typename T>
 struct RadixContainer {
   std::shared_ptr<Partition<T>> elements;
   std::vector<size_t> partition_offsets;
+
+  // bit vector to store NULL flags
+  std::shared_ptr<std::vector<bool>> position_is_null_value;
 };
 
-template <typename T, typename HashedType>
+inline std::vector<size_t> determine_chunk_offsets(std::shared_ptr<const Table> table) {
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: Why inline?

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 5]
```diff
@@ -70,15 +68,35 @@ template <typename T>
...
+  const size_t chunk_count = table->chunk_count();
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: auto?

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 6]
```diff
@@ -70,15 +68,35 @@ template <typename T>
...
+template <typename T, typename HashedType, bool consider_null_values = false>
 RadixContainer<T> materialize_input(const std::shared_ptr<const Table>& in_table, ColumnID column_id,
-                                    std::vector<std::vector<size_t>>& histograms, const size_t radix_bits,
-                                    bool keep_nulls = false) {
+                                    std::vector<std::vector<size_t>>& histograms, const size_t radix_bits) {
+  const std::hash<HashedType> hash_function;
   // list of all elements that will be partitioned
   auto elements = std::make_shared<Partition<T>>(in_table->row_count());
 
+  [[maybe_unused]] auto position_is_null_value = std::make_shared<std::vector<bool>>();
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: If you have a vector, it's more than a single position, right? Maybe `null_values` as in other places?

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 7]
```diff
@@ -110,36 +128,47 @@ RadixContainer<T> materialize_input(const std::shared_ptr<const Table>& in_table
...
             } else {
-              *(output_iterator++) =
-                  PartitionedElement<T>{RowID{chunk_id, value.chunk_offset()}, hashed_value, value.value()};
+              *(output_iterator++) = PartitionedElement<T>{RowID{chunk_id, value.chunk_offset()}, value.value()};
+            }
+
+            // In case we care about NULL values, store the NULL flag
+            if constexpr (consider_null_values) {
+              if (value.is_null()) {
+                *(null_vector_iterator) = true;
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: No need for parentheses.

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 8]
```diff
@@ -110,36 +128,47 @@ RadixContainer<T> materialize_input(const std::shared_ptr<const Table>& in_table
...
+              }
             }
 
             const Hash radix = hashed_value & mask;
             ++histogram[radix];
+            ++null_vector_iterator;
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: iterator should have the same name as its original object plus `_iter(ator)`

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 9]
```diff
@@ -222,13 +251,21 @@ std::vector<std::optional<HashTable<HashedType>>> build(const RadixContainer<Lef
   return hashtables;
 }
 
-template <typename T>
+template <typename T, typename HashedType, bool consider_null_values = false>
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: see above - I'd avoid the default parameter (even more so as the default is set to something that will break things if you don't set it)

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 10]
```diff
@@ -222,13 +251,21 @@ std::vector<std::optional<HashTable<HashedType>>> build(const RadixContainer<Lef
...
+    DebugAssert(radix_container.position_is_null_value->size() == radix_container.elements->size(),
+                "partition_radix_parallel() called with NULL consideration but radix container does not store any NULL "
+                "value information");
+  }
+
+  const std::hash<HashedType> hash_function;
+
   // materialized items of radix container
   const Partition<T>& container_elements = *radix_container.elements;
+  [[maybe_unused]] const std::vector<bool>& container_null_values = *radix_container.position_is_null_value;
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: auto?

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 11]
```diff
@@ -352,36 +411,65 @@ void probe(const RadixContainer<RightType>& radix_container,
           if (rows_iter != hashtable.end()) {
             // Key exists, thus we have at least one hit
             const auto& matching_rows = rows_iter->second;
+
+            // Since we cannot store NULL values directly in off-the-shelf containers,
+            // We need to the check the NULL bit vector here because a NULL value (represented
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: "we"

**Comment by Bouncner** [comment]
Performance: pretty much no changes except from two queries.

TPC-H Q9 is supposed to be a join-heavy query if I remember correctly. What should have changed in Q22 is not clear to me.

File: src/lib/operators/join_hash.cpp (hunk 1 above)
Comment (Bouncner) [reply src/lib/operators/join_hash.cpp]: Why should they have to be constexpr? 

Do we talk about the same? https://godbolt.org/z/1e9R8w

File: src/lib/operators/join_hash/join_hash_steps.hpp (hunk 4 above)
Comment (Bouncner) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: To avoid duplicate symbol errors when libhyrise and the test include the header file (there is no cpp file).

File: src/lib/operators/join_hash/join_hash_steps.hpp (hunk 3 above)
Comment (Bouncner) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: Will do.

File: src/lib/operators/join_hash.cpp (hunk 1 above)
Comment (mrks) [reply src/lib/operators/join_hash.cpp]: At least known at compile-time.

```
int main(int argc, char **argv) {
    const bool x = argc > 1;
    foo<x>();
}
```
does not work

File: src/lib/operators/join_hash/join_hash_steps.hpp (hunk 4 above)
Comment (mrks) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: Anonymous namespace?

Comment (Bouncner) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: I have no preferences apart from other functions using `inline` already.

File: src/lib/operators/join_hash.cpp (hunk 1 above)
Comment (Bouncner) [reply src/lib/operators/join_hash.cpp]: Ah ok, I see.

So simply remove the comment or is there an alternative to implement?

Comment (mrks) [reply src/lib/operators/join_hash.cpp]: Not that I am aware of. I'd love it if it would work the way you're suggesting, but I don't think the standard supports it

Comment (mrks) [reply src/lib/operators/join_hash.cpp]: read: remove the comment

File: src/lib/operators/join_hash/join_hash_steps.hpp (hunk 4 above)
Comment (mrks) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: Ah, nevermind. In a header, anonymous namespaces would violate the ODR.

**Comment by Bouncner** [comment]
So for documentation reasons here the results which changes affected the compile times and performance how much.

Compile times:
```
Before: 27s   src/lib/operators/join_hash.cpp
After:  37s   src/lib/operators/join_hash.cpp
```

Micro benchmarks:
```
Initial >>
BM_Join_SmallAndSmall<JoinHash>_median        123,452 ns
BM_Join_SmallAndBig<JoinHash>_median      322,140,684 ns
BM_Join_MediumAndMedium<JoinHash>_median   92,972,600 ns
Hash no longer stored >>
BM_Join_SmallAndSmall<JoinHash>_median        126,838 ns
BM_Join_SmallAndBig<JoinHash>_median      269,045,337 ns
BM_Join_MediumAndMedium<JoinHash>_median   71,167,068 ns
NULLability as template parameter >>
BM_Join_SmallAndSmall<JoinHash>_median        122,362 ns
BM_Join_SmallAndBig<JoinHash>_median      253,440,072 ns
BM_Join_MediumAndMedium<JoinHash>_median   70,601,544 ns
```

So for small hash joins (looking at you, OLTP) avoiding the NULL handling checks bringt some advantages. For larger joins, no longer storing the hash has a nice impact.
